default/python/AI/colonization/calculate_planet_colonization_rating.py

```diff
@@ -1404,17 +1404,20 @@ def _adapt_supply_factor(supply_factor: float, species: fo.species) -> float:
 
 def _colony_upkeep(num_exobots: int, num_normal: int, num_outpost: int) -> float:
```

Comment:
Let's add some tests for that function.   The purpose of this test is to document that method.  The current formula is a bit complicated and I cannot draw its plot I'm my head. 


Adding tests is a pretty easy process.

Add one more function, this one should be pure, just calculation on arguments.  
```
def  _calculate_colony_upkeep(
    num_exobots,
    num_normal, 
    num_outpost,  
    ip_per_colony, 
    owned_planets_at_max_upkeep,
    outpost_factor,
):
"""
Calculate colony output.

See sample values in ...<test_file.py>...

This formula should be aligned with .....
""" 
 ```

The current function will read values reference and pass them as arguments


For tests, you could use parametrization. Like here: https://github.com/freeorion/freeorion/blob/aecdf86a104b5ab14d38e64a6599ebb42b630ceb/default/python/tests/AI/test_bonus_calculator.py#L5
And do not forget to use [pytest,approx](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-approx).


Test cases: 
We could have one set of use-cases for default value refs:
- starting empire
- medium empire
- large empire
- an empire that could not grow anymore

These are bit vague terms :(
Maybe add  tests with different proportions for colonies/outposts/exobots

You could try to get some arguments from real AI games.

Reply:
@Cjkjvfnby I'm trying to follow your suggestions but I'm having a hard time. Could you help me out fixing the remaining lint errors? I don't really know what I am doing.

The upkeep equation works OK though.

Reply:
@Cjkjvfnby looks like I finally figured it out (how to read lint errors and not fall into despair).

This is ready for review and eventual merge.